Add `dtype-i128` feature flag and `Int128Type`

plaflamme opened the pull request:
This adds a new datatype: `Int128Type` which is the primitive type used by arrow to represent `Decimal` types. This new type is behind a `dtype-i128` feature flag.

Introducing this type requires abstracting over the arithmetic used for `PrimitiveArray` because decimal arithmetic must keep track of changes to `precision` / `scale`. This PR adds a new trait `ArrayArithmetics` which delegates to `basic` for all primitive types except `i128` and `decimal` for `i128`. There's probably a better approach, but this is the extent of what my Rust-foo is capable of delivering.

There are a few problems:

* `div_scalar` in [`basic`](https://docs.rs/arrow2/0.15.0/arrow2/compute/arithmetics/basic/fn.div_scalar.html) and [`decimal`](https://docs.rs/arrow2/0.15.0/arrow2/compute/arithmetics/decimal/fn.div_scalar.html) do not equivalent signatures: unclear how we can abstract over this
* `rem` and `rem_scalar` are not implemented in `decimal` (this doesn't seem like a blocker)

This is obviously just a draft to get some feedback.

ritchie46 commented on polars/polars-core/src/datatypes/mod.rs:
```diff
@@ -82,6 +82,8 @@ impl_polars_datatype!(Int8Type, Int8, i8);
 impl_polars_datatype!(Int32Type, Int32, i32);
 impl_polars_datatype!(Int64Type, Int64, i64);
+#[cfg(feature = "dtype-i128")]
+impl_polars_datatype!(Int128Type, Unknown, i128);
```
`Unknown` should be `Int128Type`

plaflamme replied:
Presumably, you mean `Int128` (which would be `DataType::Int128`)?

If so, I don't think that will work because I think `DataType` will eventually be something like `DataType::Decimal(precision, scale)`, right?

ritchie46 replied:
Yes, but then I think we should make the `DataType` something like `DataType::Decimal(Option<precision, scale>)`. Then we can fill in the black later.

ritchie46 reviewed (commented):
Thanks for the PR. I think we should not think of `decimal` yet and just implement `i128` as its own type.

The future decimal type will be a wrapper around this and that type will have to figure out how to deal with arithmetic, it is not a concern of the physical `i128` type. So the arithmetic can continue as is.

plaflamme commented:
@ritchie46 Thanks for the comments. The issue is that `NumericNative` has a trait bound of `NativeArithmetics` [here](https://github.com/pola-rs/polars/blob/master/polars/polars-core/src/datatypes/mod.rs#L178). But that's not implemented for `i128` precisely because the arithmetics for `i128` are not the same as the native `i128` Rust type (because it also needs to track precision and scale changes). This is mentioned [here](https://github.com/jorgecarleitao/arrow2/blob/ca14ef029580557e8da09726f340cd20f9a443d0/src/compute/arithmetics/basic/mod.rs#L31-L32).

So when we introduce `i128` as a polars "primitive" type, we can no longer respect this trait bound and must already abstract over it.

Unless what you're suggesting is that we don't use `i128` and use a wrapper type so we can implement `NativeArithmetics` for it?

plaflamme commented:
@ritchie46 Any additional feedback or guidance you could provide? I might have some more time over the next few days to look at this.

ritchie46 reviewed (commented):
Could you make clippy happy, then we can go forward.

plaflamme commented:
Alrighty, I've fixed the warning in https://github.com/pola-rs/polars/pull/6374/commits/b803b6db5fdcdf8497f5f51ea47c64ac00508531 and added `DataType::Decimal128` in https://github.com/pola-rs/polars/pull/6374/commits/0d0a89825891aa88948d30ec1be29bdde9277451 PTAL

ritchie46 commented:
Thanks @plaflamme. The next step is adding this `Series`. Would you want to have this merged first and then continue on the `Series` later?

plaflamme commented:
If we can get this merged, that would be preferable, yeah. Though I'd be worried with the `todo`s introduced since those will cause panics which isn't particularly good UX!

ritchie46 commented:
> If we can get this merged, that would be preferable, yeah. Though I'd be worried with the `todo!()`s introduced since those will cause panics which isn't particularly good UX!

Nope, it is an incremental effort.

ritchie46 commented:
Thanks @plaflamme. Next up a `Series` of `Decimal` :rocket: ;)

plaflamme commented:
@ritchie46 great, thanks for merging this. Do you have any pointers to get me started on `Series`? I took a very brief look and it seems like there are a few things that will need attention. Would you suggest starting with the `implementations` module?

ritchie46 commented:
Yes.. And you need to make newtype that is `ChunkedDecimal` wrapped around `ChunkedArray<128>`

plaflamme commented:
@ritchie46 am I on the right track with [this](https://github.com/plaflamme/polars/commit/bd7ed406fe5186edce6088a65d231832eeb54683)?

ritchie46 commented:
Yes! Most methods can be simply dispatched to ChunkedArray<128> from the Series implementation.

If you open a PR we could track this.

plaflamme commented:
@ritchie46 here's a PR: https://github.com/pola-rs/polars/pull/6374

aldanor commented:
@plaflamme Thanks for your work on this! I'm very exicted in seeing the int128 support becoming full-fledged (and yet another thing to distance polars from pandas...).

@ritchie46 wonder what's the approximate roadmap to getting decimals supported all the way through to the Python layer, what are the next steps from here? (I could try to contribute myself if the tasks are reasonably narrow-scoped and if it helps to speed it up)

plaflamme commented:
@aldanor I've started some work in [this branch](https://github.com/plaflamme/polars/tree/decimal-support). I don't really know what I'm doing, I'm mostly following breadcrumbs from datetime and other types that are similar to `Decimal`. In any case, help on this would be appreciated, feel free to make PRs against my branch.

@ritchie46 any guidance would be appreciated

aldanor commented:
Actually, one (perhaps weird) question to discuss before we hack too deep: could we just get rid of `precision` altogether? This would potentially simplify quite a lot of the code.

It won't change any computational logic or affect any casting as it's used solely for validation. Can't be just use max-precision, i.e. 38, internally?

Off top of my head, the only thing it can visibly affect is formatting, i.e. 123.45 with (5, 2) will get printed as "123.45" whereas 123.45 with (7, 3) will get printed as 0123.450, IIRC.
